### ISSUE TITLE
Long strings added to cognito:group array in Pre-Token Generation causes login error

### ISSUE THREAD

**Issue by benjeater** [aws-amplify/amplify-js#9695]

**Comment by benjeater** [comment]
@chrisbonifacio Has any progress been made on this issue?

**Comment by benjeater** [comment]
**TL;DR:**  CloudFront doesn't allow headers larger than 8kb, so large JWT payloads in the Cognito tokens cause rejected calls.

### Cause

Well I seem to have figured this out.  It's a limit on the header size that is accepted by CloudFront.
```
<Error>
    <Code>RequestHeaderSectionTooLarge</Code>
    <Message>Your request header section exceeds the maximum allowed size.</Message>
    <MaxSizeAllowed>8192</MaxSizeAllowed>
    <RequestId>DZE924D0XHJVGBBW</RequestId>
    <HostId>nME0WtBh3Xxl0kK3LxV0hZ+xf9v4d3jZP8ppvn19LiPRmhg9B8CHPC4v7je5SJJfSZ8XBopubVs=</HostId>
</Error>
```

https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/cloudfront-limits.html#limits-general

Because all three tokens (auth, identity, refresh) are included in the header calls after login and all `cognito:group` entries are present in all three tokens, the header size can balloon very quickly to exceed 8192 bytes.

The reason for the large number of `cognito:group` entries was to create multi-tenancy which included granular control of employee permissions (e.g. `RAW_DATA_FULL_CONTROL_620c374f-447c-411f-b606-df79f035a700` would allow the user to create, read, update, and delete entries in the `RawData` DynamoDB table for the tenantId of `620c374f-447c-411f-b606-df79f035a700`. `RULE_READ_620c374f-447c-411f-b606-df79f035a700` would allow the user to read the data cleanup rules set up for the tenant they are connected to.

Unfortunately, with a limit of about 25 entries in the `cognito:group` array, this isn't possible because I have about 50 separate granular function permissions with more planned.

### Proposed Solution

I've been looking for a few days now and no one seems to have a good solution for multi-tenancy using Amplify GraphQL API.

Proposal #5 in [#1043](https://github.com/aws-amplify/amplify-cli/issues/1043) looks interesting, but I'm not sure it solves my particular issue.  

Custom auth also doesn't solve the issue because it only runs for the whole GraphQL request, not on every nested model in the request.  This causes a massive overhead as a lambda has to check all the nested data permissions accessed in the query/mutation.
